src/sabre_swap/mod.rs

```diff
@@ -250,158 +241,81 @@ fn swap_map_trial(
-                }
-                for edge in dag.dag.edges(node) {
-                    let successor = edge.target().index();
-                    required_predecessors[successor] -= 1;
-                    if required_predecessors[successor] == 0 {
-                        front_layer.push(edge.target());
-                    }
-                }
+        // If we exceeded the number of allowed attempts without successfully routing a node, we
+        // reset back to the state we were in last time we routed a node, then find the node in the
+        // front layer whose qubits are the closest in the coupling map, and greedily insert swaps
+        // to make the node routable.  We could clone the layout each time we route a gate, but
+        // this path is only an escape mechansim for the algorithm getting stuck, so it should
+        // ideally never be taken, and it doesn't matter if it's not the speediest---it's better to
+        // keep the other path faster.
+        if routable_nodes.is_empty() {
+            undo_swaps(&mut current_swaps, &mut layout);
+            let (node, qubits) = closest_operation(&front_layer, &layout, dist);
+            swaps_to_route(&mut current_swaps, &qubits, &layout, coupling_graph);
+            for &[a, b] in current_swaps.iter() {
+                layout.swap_logical(a, b);
             }
-            qubits_decay.fill_with(|| 1.);
-            extended_set = None;
-            continue;
+            routable_nodes.push(node);
         }
-        let first_layer: Vec<[usize; 2]> = front_layer
-            .iter()
-            .map(|n| {
-                let node_weight = dag.dag.node_weight(*n).unwrap();
-                let qargs = &node_weight.1;
-                [qargs[0], qargs[1]]
-            })
-            .collect();
-        if extended_set.is_none() {
-            extended_set = Some(obtain_extended_set(
-                dag,
-                &front_layer,
-                &mut required_predecessors,
-            ));
-        }
-
-        let best_swap = sabre_score_heuristic(
-            &first_layer,
-            &mut layout,
-            neighbor_table,
-            extended_set.as_ref().unwrap(),
-            dist,
-            &qubits_decay,
-            heuristic,
-            &mut rng,
+        update_route(
+            &routable_nodes,
+            current_swaps,
+            dag,
+            &layout,
+            coupling_graph,
+            &mut gate_order,
+            &mut out_map,
+            &mut front_layer,
+            &mut extended_set,
+            &mut required_predecessors,
         );
-        num_search_steps += 1;
-        if num_search_steps >= DECAY_RESET_INTERVAL {
-            qubits_decay.fill_with(|| 1.);
-            num_search_steps = 0;
-        } else {
-            qubits_decay[best_swap[0]] += DECAY_RATE;
-            qubits_decay[best_swap[1]] += DECAY_RATE;
-        }
-        ops_since_progress.push(best_swap);
+        qubits_decay.fill(1.);
```

Comment:
Fwiw, I think there was a reason I used `fill_with()` instead of `fill()` for this (probably some random micro-optimization to save a few nanoseconds that I read about somewhere a while ago) but I can't remember exactly why and this will work fine (and if there was a performance difference it will be imperceivable).

Reply:
tbh, the code went through a few iterations, and I'd probably lost that you'd originally used `fill_with` when I wrote the `fill` version.  As long as there's no meaningful difference, I guess it doesn't matter (and for types that implement `Copy`, I'd be surprised if `fill_with` could be more efficient?).